test/versions_test.cr

```diff
@@ -0,0 +1,149 @@
+require "./test_helper"
+
+module Shards
+  class VersionsTest < Minitest::Test
+    def test_compare
+      # a is older than b:
+      assert_equal 1, Versions.compare("1.0.0", "1.0.1")
+      assert_equal 1, Versions.compare("1.0.0", "2.0.0")
+      assert_equal 1, Versions.compare("1.0", "1.0.0.1")
+      assert_equal 1, Versions.compare("1.0.0", "1.0.0.1")
+
+      # a == b
+      assert_equal 0, Versions.compare("0.1", "0.1")
+      assert_equal 0, Versions.compare("0.1", "0.1.0.0")
+      assert_equal 0, Versions.compare("0.1.0", "0.1")
+      assert_equal 0, Versions.compare("2.0.0", "2.0.0")
+
+      # a is newer than b:
+      assert_equal -1, Versions.compare("1.0.1", "1.0.0")
+      assert_equal -1, Versions.compare("2.0.0", "1.0.0")
+      assert_equal -1, Versions.compare("1.0.0.1", "1.0")
+      assert_equal -1, Versions.compare("1.0.0.1", "1.0.0")
+    end
+
+    def test_compare_preversions
+      # a is older than b:
+      assert_equal 1, Versions.compare("1.0.0-beta", "1.0.0")
+      assert_equal 1, Versions.compare("1.0.0.alpha", "1.0.0")
+      assert_equal 1, Versions.compare("1.0.0.alpha", "1.0.0.beta")
+      assert_equal 1, Versions.compare("1.0.beta", "1.0.0")
+      assert_equal 1, Versions.compare("1.0.alpha", "1.0.0-beta")
+
+      assert_equal 1, Versions.compare("1.0-pre1", "1.0-pre2")
+      assert_equal 1, Versions.compare("1.0-pre2", "1.0-pre10")
+
+      # a == b
+      assert_equal 0, Versions.compare("1.0.0-beta", "1.0.0-beta")
+      assert_equal 0, Versions.compare("1.0.0-alpha", "1.0.0.alpha")
+      assert_equal 0, Versions.compare("1.0.beta", "1.0.0.beta")
+      assert_equal 0, Versions.compare("1.0.beta", "1.0.0.0.0.0.beta")
+
+      # a is newer than b:
+      assert_equal -1, Versions.compare("1.0.0", "1.0.0-beta")
+      assert_equal -1, Versions.compare("1.0.0", "1.0.0.alpha")
+      assert_equal -1, Versions.compare("1.0.0.beta", "1.0.0.alpha")
+      assert_equal -1, Versions.compare("1.0.0", "1.0.beta")
+      assert_equal -1, Versions.compare("1.0.0-beta", "1.0.alpha")
+
+      assert_equal -1, Versions.compare("1.0-pre2", "1.0-pre1")
+      assert_equal -1, Versions.compare("1.0-pre10", "1.0-pre2")
+    end
+
+    def test_sort
+      100.times do
```

Comment:
Why so many randomized iterations? The comparisons between items always stay the same, so this doesn't provide any additional coverage.
You don't get any more complete result than iterating through the sorted list and asserting that every item compares lower than its successors.

Reply:
The original NaturalSort had a bug where the initial order influenced the output ordering. Adding randomization catched it. 100 is a magic number, chances of not catching the issue is veeery low, and this has no impact on speed test (fast).

Reply:
It just has some smell to use randomized brute force specs 😄